# **Transformer**
## 1. Transformer가 등장한 이유

|기존 구조|한계점|해결책 (Transformer)|
|---|---|---|
|**RNN / LSTM**|단어를 순서대로 처리 → 병렬 학습 불가|Attention 구조로 문맥 병렬 이해|
|**CNN**|지역 단어 관계만 학습|전 문맥(Global context) 학습 가능|
|**Transformer**|전체 문장을 동시에 입력받아, 단어 간의 **의미 관계**를 학습|문맥 이해 + 빠른 학습 + 대규모 데이터 처리|

> 💬 한 문장 요약
> 
> “Transformer는 **모든 단어가 서로의 의미를 바라보는(attend)** 구조다.”

---

## ⚙️ 2. Transformer의 핵심 구조
![[image/Pasted image 20251023090919.png]]
![[image/Pasted image 20251023091127.png]]
### 전체 구조 개요

```
입력 문장 → 임베딩 → Self-Attention → Feed Forward → (반복 N회)

```

|구성요소|설명|
|---|---|
|**Embedding**|단어를 벡터(숫자)로 표현|
|**Positional Encoding**|단어의 순서 정보를 더함|
|**Self-Attention**|단어 간의 관계를 학습|
|**Feed Forward**|비선형 변환으로 문맥 정제|
|**Residual + LayerNorm**|정보 손실 방지 및 안정적 학습|

---

## 🔍 3. Self-Attention 이해하기 (GPT의 핵심)

> 단어 간의 “관계 강도”를 계산하는 메커니즘
> 
> 즉, 한 문장 내 단어들이 서로 **얼마나 관련 있는지**를 수치화함

### 예시 문장

> “The cat sat on the mat.”

|단어|관계가 강하게 연결되는 단어|
|---|---|
|cat|sat, mat|
|mat|on, cat|
|the|거의 모든 단어와 약한 연결|

### 💡 핵심 아이디어

- 모델은 “cat”을 볼 때 “sat”, “mat”을 **같이 참조(attend)** 하여 문맥을 이해
- 이런 방식으로 “순서”가 아닌 “의미 관계” 중심의 학습이 가능

In [1]:
from transformers import AutoTokenizer
tok = AutoTokenizer.from_pretrained("gpt2")

text = "Deep learning is amazing"
tokens = tok.tokenize(text)
print("🔹 Tokenized:", tokens)
print("🔹 Token IDs:", tok.convert_tokens_to_ids(tokens))

🔹 Tokenized: ['Deep', 'Ġlearning', 'Ġis', 'Ġamazing']
🔹 Token IDs: [29744, 4673, 318, 4998]


🔹 Tokenized: ['Deep', 'Ġlearning', 'Ġis', 'Ġamazing']
🔹 Token IDs: [29744, 4673, 318, 4998]

> Ġ = 공백을 의미   
> 문맥 수준의 의미 파악

In [2]:
from transformers import pipeline

# Hugging Face의 사전 학습된 모델 사용
# (번역 / 문장 생성 / 요약 등 다양한 태스크)
translator = pipeline("translation", model="Helsinki-NLP/opus-mt-tc-big-en-ko")
generator = pipeline("text-generation", model="gpt2")
summarizer = pipeline("summarization", model="facebook/bart-large-cnn")

print("✅ 준비 완료! 이제 바로 실습 시작합니다.")


Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


source.spm:   0%|          | 0.00/790k [00:00<?, ?B/s]

c:\Users\shlim\miniconda3\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\shlim\.cache\huggingface\hub\models--Helsinki-NLP--opus-mt-tc-big-en-ko. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP

target.spm:   0%|          | 0.00/815k [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/65.0 [00:00<?, ?B/s]

c:\Users\shlim\miniconda3\Lib\site-packages\transformers\models\marian\tokenization_marian.py:175: UserWarning: Recommended: pip install sacremoses.
  warnings.warn("Recommended: pip install sacremoses.")
Device set to use cuda:0
Device set to use cuda:0


config.json: 0.00B [00:00, ?B/s]

c:\Users\shlim\miniconda3\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\shlim\.cache\huggingface\hub\models--facebook--bart-large-cnn. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download.

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

Device set to use cuda:0


✅ 준비 완료! 이제 바로 실습 시작합니다.


In [3]:
english_sentences = [
    "Machine learning is transforming the way we live.",
    "Artificial intelligence can help doctors diagnose diseases faster.",
    "ChatGPT is changing how people learn and work."
]

for s in english_sentences:
    result = translator(s)[0]['translation_text']
    print(f"🔹 English: {s}\n🔸 Korean : {result}\n")


🔹 English: Machine learning is transforming the way we live.
🔸 Korean : Kei popular 잘 NBA 유니버설 안수.

🔹 English: Artificial intelligence can help doctors diagnose diseases faster.
🔸 Korean : Greece COMP은 그리스의 도시입니다.

🔹 English: ChatGPT is changing how people learn and work.
🔸 Korean : 06:02 그때 springs 대중적인 pleasingtruepatriot 끝 세그먼트.



In [5]:
prompt = "Rewrite the following sentence in a more formal way: 'AI is super cool and fun to use.'"
result = generator(prompt, max_length=40, num_return_sequences=1)[0]['generated_text']
print("🧩 Result:\n", result)


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.
Both `max_new_tokens` (=256) and `max_length`(=40) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


🧩 Result:
 Rewrite the following sentence in a more formal way: 'AI is super cool and fun to use.' In fact, it's not. Instead, it's about fun.

This isn't to say that AI is not fun, although it is certainly fun to use.

AI is a very good way to go about your job.

Why AI?

AI is a very good way to go about your job. To put it simply, it's about fun.

Because it's a good way to go about your job, and because it's a great way to go about your job. To put it simply, it's about fun.

The problem with AI is that it's not very good at it. (In fact, in some cases, AI is often actually quite good at it.) It takes a lot of effort to make it work, and it's hard to use it to your advantage.

But if you're not able to put much effort into making your job work, why not use the AI to your advantage? Here's why:

It takes less time to learn the concepts of computation, and is much easier to understand, and to use.

It takes less time to learn the concepts of computation, and is much easier to underst

In [6]:
text = """
Artificial intelligence (AI) refers to the simulation of human intelligence
in machines that are programmed to think like humans and mimic their actions.
The term may also be applied to any machine that exhibits traits
associated with a human mind such as learning and problem-solving.
"""

# Step 1: 영어 요약 (GPT 기반 Summarization)
summary = summarizer(text, max_length=50, min_length=25, do_sample=False)[0]['summary_text']
print("🔹 English Summary:\n", summary, "\n")

# Step 2: 요약된 문장을 한국어로 번역
translation = translator(summary)[0]['translation_text']
print("🔸 Korean Translation:\n", translation)


🔹 English Summary:
 Artificial intelligence (AI) refers to machines that are programmed to think like humans and mimic their actions. The term may also be applied to any machine that exhibits traits associated with a human mind. 

🔸 Korean Translation:
 은퇴                                                                                .                 .                                                .
